camel/memories/blocks/graphdb_block.py

```diff
@@ -0,0 +1,176 @@
+        database: Optional[str] = "neo4j",
+    ) -> None:
+        self.storage = storage or self.default_graph_storage(
+            url, username, password, database
+        )
+
+    def default_graph_storage(
+        self,
+        url: Optional[str] = None,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        database: Optional[str] = "neo4j",
+    ) -> BaseGraphStorage:
+        # Try to get details from environment variables first
+        url = os.getenv('NEO4J_URI', url)
+        username = os.getenv('NEO4J_USERNAME', username)
+        password = os.getenv('NEO4J_PASSWORD', password)
+        database = os.getenv('NEO4J_DATABASE', database)
+
+        # Log an error if any of the necessary values are missing
+        if not url or not username or not password:
+            logger.error(
+                "Neo4j connection detailed are missing."
+                "Ensure environment variables or parameters are set for"
+                "NEO4J_URI, NEO4J_USERNAME, and NEO4J_PASSWORD."
+            )
+            raise ValueError("Missing Neo4j connection details.")
+
+        # Create and return an instance of Neo4jGraph
+        return Neo4jGraph(url, username, password, database or "neo4j")
+
+    def retrieve(
+        self, query: str, params: Optional[Dict[str, Any]] = None
+    ) -> List[ContextRecord]:
+        r"""Retrieves records from the graph database based on a query.
+
+        Args:
+            query (str): The query string to execute.
+            params (Optional[Dict[str, Any]]): Parameters for the query.
+
+        Returns:
+            List[ContextRecord]: A list of context records retrieved from the
+                graph database.
+        """
+        try:
+            results = self.storage.query(query, params)
+            scores = self._calculate_scores(query, results)
+            return [
+                ContextRecord(
+                    memory_record=MemoryRecord.from_dict(result), score=score
+                )
+                for result, score in zip(results, scores)
+            ]
+        except Exception as e:
+            logger.error(f"Error retrieving records: {e}")
+            return []
+
+    def write_records(self, records: List[MemoryRecord]) -> None:
+        """Writes records to the graph database."""
+        for record in records:
+            content = record.message.content
+
+            if isinstance(content, str):
+                try:
+                    data = json.loads(content)
+                    subj = data.get('subject')
+                    obj = data.get('object')
+                    rel = data.get('relation')
+                except json.JSONDecodeError:
+                    logger.error(f"Error parsing content as JSON: {content}")
+                    continue
+            else:
+                subj = content.get('subject')
+                obj = content.get('object')
+                rel = content.get('relation')
```

Comment:
`record.message.content` should be `str`, for what case it will go to this part?